Add InBandMessagesSince to RPC and NetworkInterfaceIncoming

jacobhaven opened the pull request:
r? @maxtaco @patrickxb

patrickxb commented on protocol/gregor1/avdl/incoming.avdl:
```diff
@@ -3,4 +3,5 @@
 
 protocol incoming {
 	void consumeMessage(Message m);
+	array<InBandMessage> inBandMessagesSince(UID u, DeviceID d, Time t);
```
u, d, and t seem a little terse for protocol parameters...

jacobhaven replied:
Fair, *updated*

jacobhaven commented:
It's a bit annoying that we have `gregor1.IncomingInterface` and `gregor.NetworkInterfaceIncoming`, which are basically the same, except for the protocol/interface type of the args. Should we consolidate onto just one? (I don't particularly care which)

maxtaco commented:
Happy with consolidating.  We have the same issue in some places in the main client too

maxtaco commented on protocol/gregor1/avdl/incoming.avdl:
```diff
@@ -3,4 +3,5 @@
 
 protocol incoming {
 	void consumeMessage(Message m);
+	array<InBandMessage> inBandMessagesSince(UID uid, DeviceID deviceid, Time ctime);
```
`TimeOrOffset` is better for clients that don't trust their clock

jacobhaven replied:
They should request from the last known `CTime` (that was set for them by the server). If they don't have that, they should request everything (send `ctime=0`)

jacobhaven commented:
I just uncoupled a separate `sync.avdl`/`SyncInterface` to avoid touching `gregor.NetworkInterfaceIncoming`

maxtaco commented on storage/sql_sm.go:
```diff
@@ -483,6 +502,26 @@ func (s *SQLEngine) InBandMessagesSince(u gregor.UID, d gregor.DeviceID, t grego
+		arg.Deviceid = gregor1.DeviceID(d.Bytes())
+	}
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
+	newMsgs, err := s.sync.InBandMessagesSince(ctx, arg)
```
why this?

jacobhaven replied:
? It's not recursive. It calls out an RPC to get the messages not stored in the `StateMachine` (appended to the ones from the `StateMachine`)

maxtaco commented:
I think this is going in the right direction. Can you add some tests? Also can you assert after the catch up that you have the same state that the server does?

jacobhaven commented:
I've refactored a lot of things.
* Renamed the RPC to just `Sync` instead of `InBandMessagesSince` to avoid confusion with the `StateMachine` method.
* Removed Networking Interfaces from "gregor", so they could be handled entirely by the rpc protocol "gregor1"
* Added the server-side plumbing for `Sync` through `connection` and `Server` 
* Fleshed out client tests, checking hashes of State against the server and restored client.

jacobhaven commented on storage/sql_sm.go:
```diff
@@ -462,7 +462,7 @@ func (s *SQLEngine) InBandMessagesSince(u gregor.UID, d gregor.DeviceID, t time.
 	}
 
-	qb.Build("AND m.ctime >= ")
+	qb.Build("AND m.ctime > ")
```
I'm pretty sure these are the semantics we want. Otherwise we end up duplicating the last message (a similar change was made in `MemEngine`)

maxtaco replied:
I think it was right as is, since you could get two messages in the same microsecond, but you only saw one and then you hung up.  The protocol should be idempotent / resilient to repeated messages.

In fact, I was thinking, just to make sure, we'd ask for messages since (last-ctime - 1 minute) or so, since messages can be reordered, and we might have missed some late messages as we were shutting down.

maxtaco commented on storage/client.go:
```diff
@@ -62,3 +68,40 @@ func (c *Client) Restore() error {
+	arg := gregor1.SyncArg{
+		Uid:      gregor1.UID(c.user.Bytes()),
+		Deviceid: gregor1.DeviceID(c.device.Bytes()),
+		Ctime:    gregor1.ToTime(ctime),
```
what to you think about `ctime - time.Minute` (as per earlier comment) ?

jacobhaven replied:
I don't know what earlier comment you're referring to. Now that I've added `MsgID` deduplication code to `MemEngine`, this should work fine.

maxtaco commented on storage/sql_sm.go:
```diff
@@ -453,8 +462,8 @@ func (s *SQLEngine) InBandMessagesSince(u gregor.UID, d gregor.DeviceID, t grego
 
-	qb.Build("AND m.ctime >= ")
-	qb.TimeOrOffset(t)
+	qb.Build("AND m.ctime > ")
```
see earlier comment

maxtaco commented on storage/client_test.go:
```diff
@@ -37,4 +38,8 @@ func TestLevelDBClient(t *testing.T) {
 		t.Fatal(err)
 	}
+
+	if err := c.Sync(); err != nil {
```
does this actually get any new messages?  don't you need to generate some new ones to assert that you're using the `Sync` RPC?

jacobhaven replied:
The right thing to do here is to create two separate `StateMachine`s, give them different states and `Sync` between them.

maxtaco replied:
Does this check that the state winds up being the same across the two devices (maybe via hash?)

maxtaco commented on storage/mem_sm.go:
```diff
@@ -311,12 +312,24 @@ func (m *MemEngine) State(u gregor.UID, d gregor.DeviceID, t gregor.TimeOrOffset
-	msg := user.replayLog(m.clock.Now(), d, t)
-	return msg, nil
+	for _, u := range m.users {
+		_, ctime := u.replayLog(m.clock.Now(), nil, time.Time{})
```
Will this be expensive on mobile devices on startup, to iterate through all existing messages to find out when the last one was?

jacobhaven replied:
I can't really think of a better way to handle this without maintaining the `user` log as a list ordered by `ctime`. Maybe that is the way to go...

jacobhaven replied:
Actually, we don't need the *precise* `LatestCTime`, so we can probably get away with iterating backwards until we find the last message with the correct user/device IDs.

maxtaco commented on storage/sql_sm.go:
```diff
@@ -378,7 +378,7 @@ func (s *SQLEngine) inBandMetadataSince(u gregor.UID, t gregor.TimeOrOffset) ([]
 	qb.Build(qry, hexEnc(u))
 	if t != nil {
-		qb.Build("AND ctime >= ")
+		qb.Build("AND ctime > ")
```
Also still think `>=` is better here....

jacobhaven replied:
*reverted*

maxtaco commented on storage/sql_sm.go:
```diff
@@ -435,7 +435,16 @@ func (s *SQLEngine) rowToInBandMessage(u gregor.UID, rows *sql.Rows) (gregor.InB
-func (s *SQLEngine) InBandMessagesSince(u gregor.UID, d gregor.DeviceID, t gregor.TimeOrOffset) ([]gregor.InBandMessage, error) {
+func (s *SQLEngine) LatestCTime() *time.Time {
+	var ctime timeScanner
+	err := s.driver.QueryRow(`SELECT ctime FROM messages ORDER BY m.ctime DESC`).Scan(&ctime)
```
Can you add `LIMIT 1`?

maxtaco replied:
Actually, can you do `SELECT MAX(ctime) FROM message`?

jacobhaven replied:
Seems reasonable, *done*.

maxtaco commented on test/mocks.go:
```diff
@@ -0,0 +1,44 @@
+	return m.SM.ConsumeMessage(msg)
+}
+
+func (m *MockIncomingInterface) Sync(_ context.Context, arg gregor1.SyncArg) (res gregor1.SyncResult, err error) {
```
this mock seems very similar to the actual code.  is there any way to consolidate the two?

jacobhaven replied:
*done*

jacobhaven commented on storage/client.go:
```diff
@@ -62,3 +68,40 @@ func (c *Client) Restore() error {
+	}
+
+	if !bytes.Equal(res.Hash, hash) {
+		return errors.New("local state hash != server state hash")
```
In this case, we should discard our local state and replace it with a `Sync` from time `0` (if the hash fails after that, something is really wrong and we can error out.)

jacobhaven replied:
**Done**

maxtaco commented on storage/sql_sm.go:
```diff
@@ -435,9 +435,12 @@ func (s *SQLEngine) rowToInBandMessage(u gregor.UID, rows *sql.Rows) (gregor.InB
+func (s *SQLEngine) LatestCTime(u gregor.UID, d gregor.DeviceID) *time.Time {
+	qry := `SELECT MAX(ctime) 
+			FROM messages
+			WHERE uid = ? AND devid = ?`
```
`(devid = ? OR devid IS NULL)` since a `NULL` devid means "for all devices"

maxtaco commented on storage/mem_sm.go:
```diff
@@ -318,11 +318,18 @@ func (m *MemEngine) State(u gregor.UID, d gregor.DeviceID, t gregor.TimeOrOffset
+	log := m.getUser(u).log
+	for i := len(log) - 1; i >= 0; i-- {
+		if log[i].i != nil && log[i].i.item != nil &&
+			(d == nil || (log[i].i.item.Metadata() != nil && log[i].i.item.Metadata().DeviceID() != nil &&
```
if `log[i].i.item.Metadata().DeviceID() == nil` then this message counts too, since `nil` deviceID means "all devices"

jacobhaven replied:
**updated**

maxtaco replied:
Thanks! Lemme just take a quick read through the test before I give it a
thumb. I hope to get to it a little later tonight.

On Saturday, April 16, 2016, Jacob H. Haven <notifications@github.com>
wrote:

> In storage/mem_sm.go
> <https://github.com/keybase/gregor/pull/34#discussion_r59973463>:
>
> >  	m.Lock()
> >  	defer m.Unlock()
> > -	_, latestCTime = m.getUser(u).replayLog(m.clock.Now(), d, time.Time{})
> > -	return
> > +	log := m.getUser(u).log
> > +	for i := len(log) - 1; i >= 0; i-- {
> > +		if log[i].i != nil && log[i].i.item != nil &&
> > +			(d == nil || (log[i].i.item.Metadata() != nil && log[i].i.item.Metadata().DeviceID() != nil &&
>
> *updated*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keybase/gregor/pull/34/files/ce05afba7b91f16d0405a76aaa0272cbd55eb78c..7204ce0e1a4f23be881a8bfcfa3ba780c8c5997f#r59973463>
>

maxtaco commented on storage/sql_sm.go:
```diff
@@ -457,7 +457,7 @@ func (s *SQLEngine) Clear() error {
 	qry := `SELECT MAX(ctime) 
 			FROM messages
-			WHERE uid = ? AND devid = ?`
+			WHERE uid = ? AND devid = ? OR devid IS NULL`
```
I'm pretty sure you need parens here!  `AND` binds tighter than `OR`.

jacobhaven replied:
**fixed**

maxtaco commented on storage/client.go:
```diff
@@ -62,3 +69,55 @@ func (c *Client) Restore() error {
+	if err := c.syncFromTime(c.sm.LatestCTime(c.user, c.device)); err != nil {
+		if _, ok := err.(errHashMismatch); ok {
+			log.Printf("Sync failure: %v\nReseting StateMachine and retrying")
+			//c.sm.Clear()
```
Delete this line?

jacobhaven replied:
Oh, right. *uncommented*

maxtaco commented on storage/client_test.go:
```diff
@@ -106,6 +106,17 @@ func TestLevelDBClient(t *testing.T) {
+	if err := c.syncFromTime(c.sm.LatestCTime(c.user, c.device)); err != nil {
+		t.Fatal(err)
+	}
+
```
Can you explain why calling `AddStateMAchinePerDevice` causes `syncFromTime` to fail?

maxtaco commented:
👍
